bugshaker/src/main/java/com/github/stkent/bugshaker/email/screenshot/maps/MapBitmapObservable.java

```diff
@@ -0,0 +1,69 @@
+/**
+ * Copyright 2016 Stuart Kent
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.github.stkent.bugshaker.email.screenshot.maps;
+
+import android.graphics.Bitmap;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import com.google.android.gms.maps.GoogleMap;
+import com.google.android.gms.maps.MapView;
+import com.google.android.gms.maps.OnMapReadyCallback;
+
+import rx.Observable;
+import rx.Subscriber;
+import rx.android.schedulers.AndroidSchedulers;
+
+final class MapBitmapObservable {
+
+    static Observable<LocatedBitmap> create(@NonNull final MapView mapView) {
+        final int[] locationOnScreen = new int[] {0, 0};
+        mapView.getLocationOnScreen(locationOnScreen);
+
+        final Observable<LocatedBitmap> result = Observable.create(new Observable.OnSubscribe<LocatedBitmap>() {
+            @Override
+            public void call(final Subscriber<? super LocatedBitmap> subscriber) {
+                mapView.getMapAsync(new OnMapReadyCallback() {
+                    @Override
+                    public void onMapReady(@NonNull final GoogleMap googleMap) {
+                        googleMap.snapshot(new GoogleMap.SnapshotReadyCallback() {
+                            @Override
+                            public void onSnapshotReady(@Nullable final Bitmap bitmap) {
+                                if (bitmap != null) {
+                                    subscriber.onNext(
+                                            new LocatedBitmap(bitmap, locationOnScreen));
+
+                                    subscriber.onCompleted();
+                                } else {
+                                    subscriber.onError(new NullPointerException());
+                                }
+                            }
+                        });
+                    }
+                });
+            }
+        });
+
+        // Required, since MapView.onMapReady _must_ be called from the main thread.
+        return result.subscribeOn(AndroidSchedulers.mainThread());
```

Comment:
Based on the image at the bottom of this post, it seems that the correct structure for this situation would be as followed:

- subscribe on the io thread
- make sure the Observable fed into the `MapBitmapObservable.create` method observes on the main thread (so that all events are fed to `MapBitmapObservable.create`, which calls `GoogleMap.getMapAsync`
- make sure the `MapBitmapObservable` observes on the io thread (so that operators further up the chain will perform their manipulations on the io thread once more).

Something like:

```java
    @NonNull
    private Observable<List<LocatedBitmap>> getMapViewBitmapsObservable(@NonNull final List<MapView> mapViews) {
        return Observable
                .from(mapViews)
                .observeOn(AndroidSchedulers.mainThread())
                .concatMap(new Func1<MapView, Observable<LocatedBitmap>>() {
                    @Override
                    public Observable<LocatedBitmap> call(@NonNull final MapView mapView) {
                        return MapBitmapObservable.create(mapView);
                    }
                })
                .observeOn(Schedulers.io())
                .toList();
    }
```

It feels a little dirty to have to call `observeOn(Schedulers.io())` rather than being able to 'return' to the thread we were on implicitly before we bumped over to the main thread, so maybe there's a better way to do that...

 ![](http://reactivex.io/documentation/operators/images/schedulers.png)